pkg/e2e/verify/check_workloads.go

```diff
@@ -0,0 +1,132 @@
+
+			_, infra, err := listNodesByType(h.Kube().CoreV1())
+			Expect(err).NotTo(HaveOccurred())
+
+			_, err = checkPods(h.Kube().CoreV1(), infra)
+			Expect(err).NotTo(HaveOccurred())
+
+		})
+	})
+})
+
+func checkPods(podClient v1.CoreV1Interface, infra []string) (map[string]string, error) {
+
+	type (
+		PodName      = string
+		Namespace    = string
+		OperatorName = string
+		Node         = string
+	)
+	violators := make(map[PodName]Node)
+	operators := map[OperatorName]Namespace{"cloud-ingress-operator": "openshift-cloud-ingress-operator",
+		"configure-alertmanager-operator": "openshift-monitoring",
+		"custom-domains-operator":         "openshift-custom-domains-operator",
+		"managed-upgrade-operator":        "openshift-managed-upgrade-operator",
+		"managed-velero-operator":         "openshift-velero",
+		"must-gather-operator":            "openshift-must-gather-operator",
+		"osd-metrics-exporter":            "openshift-osd-metrics",
+		"rbac-permissions-operator":       "openshift-rbac-permissions"}
+
+	listOpts := metav1.ListOptions{}
+
+	list, err := podClient.Pods(metav1.NamespaceAll).List(context.TODO(), listOpts)
+	if err != nil {
+		return nil, fmt.Errorf("error getting pod list: %v", err)
+	}
+
+	if len(list.Items) == 0 {
+		return nil, fmt.Errorf("pod list is empty. this should NOT happen")
+	}
+
+	for _, pod := range list.Items {
+		//For each pod.ObjectMeta.Name, pod.Spec.NodeName we will check if they are a match for the operaor namespace and pod name prefix.
+		for op, namespace := range operators {
+			if pod.ObjectMeta.Namespace == namespace {
+				if (strings.HasPrefix(pod.ObjectMeta.Name, op)) && !(strings.HasPrefix(pod.ObjectMeta.Name, op+"-registry")) {
+					if !stringInSlice(pod.Spec.NodeName, infra) {
+						violators[pod.ObjectMeta.Name] = pod.Spec.NodeName
+						log.Printf(" Violation detected: pod %s, doesn't run on infra node but instead runs on %s", pod.ObjectMeta.Name, pod.Spec.NodeName)
+					}
+				}
+			}
+		}
+	}
+
+	if len(violators) > 0 {
+		return violators, fmt.Errorf("Found infrastructure pods that do not run on infra nodes %v.", err)
+	}
+
+	return violators, nil
+}
+
+func listNodesByType(nodeClient v1.CoreV1Interface) ([]string, []string, error) {
+
+	log.Printf("Getting node list")
+
+	listOpts := metav1.ListOptions{}
+	list, err := nodeClient.Nodes().List(context.TODO(), listOpts)
+	var worker []string
+	var infra []string
+	if err != nil {
+		return nil, nil, fmt.Errorf("error getting node list: %v", err)
+	}
+
+	if len(list.Items) == 0 {
+		return nil, nil, fmt.Errorf("no nodes found")
+	}
+
+	for _, node := range list.Items {
+		labels := node.Labels
+
+		for key, _ := range labels {
+			if key == "node-role.kubernetes.io/worker" {
+				worker = append(worker, node.Name)
+			}
+			if key == "node-role.kubernetes.io/infra" {
+				infra = append(infra, node.Name)
+				log.Printf("Node, infra: %v", node.Name)
+			}
+		}
+
+	}
+
+	return worker, infra, nil
+}
```

Comment:
A final suggestion: this is a way to communicate which returned slice contains which kind of nodes without significant code change. A comment would also work. Right now, the only way to figure this out is to read the source code, and I'd rather the info be surfaced another way.

```suggestion
func listNodesByType(nodeClient v1.CoreV1Interface) (worker, infra []string, err error) {

	log.Printf("Getting node list")
	var (
		listOpts metav1.ListOptions
		list *v1.NodeList
	)
	list, err = nodeClient.Nodes().List(context.TODO(), listOpts)
	if err != nil {
		return nil, nil, fmt.Errorf("error getting node list: %v", err)
	}

	if len(list.Items) == 0 {
		return nil, nil, fmt.Errorf("no nodes found")
	}

	for _, node := range list.Items {
		labels := node.Labels

		for key, _ := range labels {
			if key == "node-role.kubernetes.io/worker" {
				worker = append(worker, node.Name)
			}
			if key == "node-role.kubernetes.io/infra" {
				infra = append(infra, node.Name)
				log.Printf("Node, infra: %v", node.Name)
			}
		}

	}

	return worker, infra, nil
}
```

Reply:
I have trouble with declarations of this type  list *v1.NodeList, I get "NodeList not declared by package v1compiler" error

Reply:
The list, err := nodeClient.Nodes().List(context.TODO(), listOpts) is similar to other calls we use to list kubernetes objects, I will put some extra comments in the code to clarify.